helix-term/src/ui/explorer.rs

```diff
@@ -0,0 +1,1527 @@
+            }
+        }
+
+        if let Some((_, prompt)) = self.prompt.as_mut() {
+            prompt.render_prompt(prompt_area, surface, cx)
+        }
+    }
+
+    fn render_help(&mut self, area: Rect, surface: &mut Surface, cx: &mut Context) {
+        Info::new(
+            "Explorer",
+            &[
+                ("?", "Toggle help"),
+                ("a", "Add file/folder"),
+                ("r", "Rename file/folder"),
+                ("d", "Delete file"),
+                ("B", "Change root to parent folder"),
+                ("]", "Change root to current folder"),
+                ("[", "Go to previous root"),
+                ("+, =", "Increase size"),
+                ("-, _", "Decrease size"),
+                ("C-t", "Toggle preview (left/right only)"),
+                ("q", "Close"),
+            ]
+            .into_iter()
+            .chain(ui::tree::tree_view_help().into_iter())
+            .collect::<Vec<_>>(),
+        )
+        .render(area, surface, cx)
+    }
+
+    fn handle_prompt_event(&mut self, event: &KeyEvent, cx: &mut Context) -> EventResult {
+        fn handle_prompt_event(
+            explorer: &mut Explorer,
+            event: &KeyEvent,
+            cx: &mut Context,
+        ) -> Result<EventResult> {
+            let (action, mut prompt) = match explorer.prompt.take() {
+                Some((action, p)) => (action, p),
+                _ => return Ok(EventResult::Ignored(None)),
+            };
+            let line = prompt.line();
+
+            let current_item_path = explorer.tree.current_item()?.path.clone();
+            match (&action, event) {
+                (PromptAction::CreateFileOrFolder, key!(Enter)) => {
+                    if line.ends_with(std::path::MAIN_SEPARATOR) {
+                        explorer.new_folder(line)?
+                    } else {
+                        explorer.new_file(line)?
+                    }
+                }
+                (PromptAction::RemoveFolder, key) => {
+                    if let key!('y') = key {
+                        close_documents(current_item_path, cx)?;
+                        explorer.remove_folder()?;
+                    }
+                }
+                (PromptAction::RemoveFile, key) => {
+                    if let key!('y') = key {
+                        close_documents(current_item_path, cx)?;
+                        explorer.remove_file()?;
+                    }
+                }
+                (PromptAction::RenameFile, key!(Enter)) => {
+                    close_documents(current_item_path, cx)?;
+                    explorer.rename_current(line)?;
+                }
+                (_, key!(Esc) | ctrl!('c')) => {}
+                _ => {
+                    prompt.handle_event(&Event::Key(*event), cx);
+                    explorer.prompt = Some((action, prompt));
+                }
+            }
+            Ok(EventResult::Consumed(None))
+        }
```

Comment:
Do we need an extra layer of function here?

Reply:
This is not necessary anymore, I will remove it.